### ISSUE TITLE
VD Experimental - current known issues

### ISSUE THREAD

**Issue by zeffii** [nortikin/sverchok#2615]
- [ ] does not show dashed lines + custom fragment shader
    (disable dashed line if you need the custom fragment shader.. for now )
- [ ] does not automatically use polygon offset yet. 
    (until we can ensure that we restore opengl state when sverchok is finished drawing
- [ ] mac openGL implementation does not support glLineWidth > 1.0 
- [ ] you need to use recent version of Blender 2.8+

**Comment by Durman** [comment]
Looks like non convex polygons does not supported yet

![2019-10-19_23-00-16](https://user-images.githubusercontent.com/28003269/67150139-cb4f6280-f2c4-11e9-8441-0cf58b336b33.png)

**Comment by portnov** [comment]
Were they supported in 2.79?...

**Comment by Durman** [comment]
There was option on N panel.

**Comment by zeffii** [comment]
ah.... if the non convex is a Quad,  it should be tessellated, instead of my assumption `[[a, b, c], [a, c d]]`

**Comment by zeffii** [comment]
see.. 

https://github.com/nortikin/sverchok/blob/5cdd2a8fd35127c8066f038181dd56ff71c07121/nodes/viz/vd_draw_experimental.py#L67-L86

**Comment by zeffii** [comment]
i'll add a property, so

```python

   elif num_verts == 4 and not self.convex_quads:
         ...
```

let me consider this over night :)

**Comment by Durman** [comment]
For convex polygons it is possible to use such triangulation
```Python
if is_convex([coords[idx] for idx in idxset]):
    new_faces = []
    for i in range(2, len(idxset)):
        new_faces.append((idxset[0], idxset[i - 1], idxset[i]))
```

**Comment by zeffii** [comment]
@Durman in b28 openGL using "Core Profile", we must turn _everything_ into Triangles. The reason i "hand" tessellate the Quad is to keep the processing time of the Quad down to a minimum. Most quads we encounter are convex, and can be done this way ` a b c d -> a b c, a c d ` . It therefore doesn't make sense to always call a dedicated tessellate function on each Quad, this would penalize most polygons most of the time.

The concave quad is a special case, and testing for it first would also slow us down, so the solution will be like in VDMK2 to have a flag to say " OK, we must call the tessellate_polygon function on all non Tri faces, because we are expecting mixed input:

   1. Concave Quads, and 
   2. Irregular Ngons

that's why a test like 
```python
 elif num_verts == 4 and not self.convex_quads:
         ...
```
will result in the least speed damage to the "ensure triangles"  function.

**Comment by Durman** [comment]
As i know `is_convex` takes O(n) time. So it should not slow down process too much. And there is benefit of using such test as triangulation of convex polygons for O(n) time. 
I actually don't know how implemented the mathutils function. Perhaps it makes such test inside itself. 
I think only speed tests can lit light on this issue.

**Comment by zeffii** [comment]
i've committed the fix needed, for sake of speed vd_experimental will assume quads can be "serialized the dumb way", unless users switches on `handle_concave_quads` in the properties panel.

**Comment by Durman** [comment]
This raises cascade of errors. Weirdly.))

![2019-10-30_12-50-16](https://user-images.githubusercontent.com/28003269/67842646-fe3ef500-fb13-11e9-8b64-4a086ff73e66.png)

**Comment by zeffii** [comment]
we can add a test like 
```python

def is_input_vertex_kind(socket_data):
    return len(socket_data) and len(socket_data[0][0]) == 3

```

**Comment by Durman** [comment]
It would be nice if the node will become red.

**Comment by zeffii** [comment]
OK. that's probably reasonable.

**Comment by Durman** [comment]
Well it have taken some time for figuring out what is going wrong.

**Comment by portnov** [comment]
See also `data_structure.get_data_nesting_level`, `data_structure.ensure_nesting_level`.

**Comment by portnov** [comment]
See also `implicit_conversions` parameter of `sv_get()` method.

**Comment by zeffii** [comment]
there's a potential to add debugprint using bgl/blf directly beside the node that fails.  Like sverchok redux did (i can't find a screenshot of it):

**Comment by portnov** [comment]
I don't think it's good idea to write the whole stack trace in the viewport; maybe only the error message?

**Comment by Durman** [comment]
The same thought. Probably it would be better to make a reference to console or log file in Blender for more details.

**Comment by DolphinDream** [comment]
@zeffii Here’s an idea for fixing the linewidth > 1px issue on mac. A fix for the annotations (on mac) is presumably coming up in 2.82 (hopefully). I don’t know if this annotation issue is related to viewer draw line/edge width issue in SV,  but it’s worth checking it out.

https://developer.blender.org/T57570
https://developer.blender.org/rB014eb69cf858036816d12a4b92ffe6681978b683

**Comment by zeffii** [comment]
yes, the solution would need to be a vertex_shader / fragment_shader combo. (that's what the pull-request is ), i can experiment with it, it looks a little like the code I was experimenting with a while back.

we can make it the default behaviour for sverchok on osx, and optional for the other OS's (it might even look better because of the AA ).  

No ETA on that.

**Comment by zeffii** [comment]
i don't think i'll add https://developer.blender.org/F8403764

**Comment by zeffii** [comment]
but here's a start branch , for exactly that..  https://github.com/nortikin/sverchok/pull/2965

**Comment by zeffii** [comment]
@DolphinDream  .. it seems they removed: https://developer.blender.org/F8403764 (above link)m which is what I was basing it off.

**Comment by zeffii** [comment]
this beast: https://developer.blender.org/rB014eb69cf858036816d12a4b92ffe6681978b683

**Comment by DolphinDream** [comment]
I wonder if geometry shaders would be the way to go.. if Blender 2.8x supports them.

https://blog.tammearu.eu/posts/gllines/

**Comment by zeffii** [comment]
i think it does, but turning edges into cylinders is a little expensive :)

**Comment by zeffii** [comment]
experiment. stuff is sometimes easier than it seems, but sometimes far more complicated than it should be conceptually :)

**Comment by DolphinDream** [comment]
More interesting stuff.. with some webgl demos:

https://mattdesl.svbtle.com/drawing-lines-is-hard
https://github.com/mattdesl/webgl-lines

**Comment by zeffii** [comment]
add them here : https://github.com/nortikin/sverchok/issues/2608
